File: src/lib/models/endpoint.ts
import mongoose, { Schema, type Document } from "mongoose";

export interface IEndpoint extends Document {
  projectId: mongoose.Types.ObjectId | null;
  name: string;
  url: string;
  method: string;
  expectedStatusCode: number;
  interval: number;
  timeout: number;
  headers: Map<string, string>;
  body: string;
  isActive: boolean;
  alertEnabled: boolean;
  alertThreshold: number;
  consecutiveFailures: number;
  isAlerting: boolean;
  lastAlertedAt: Date | null;
  notifications: {
    email: { enabled: boolean; address?: string };
    discord: { enabled: boolean; webhookUrl?: string };
    webhook: { enabled: boolean; url?: string };
  };
  lastCheckedAt: Date | null;
  lastStatus: "UP" | "DOWN" | "DEGRADED" | null;
  lastResponseTime: number | null;
  totalChecks: number;
  successfulChecks: number;
  uptimePercentage: number;
  createdAt: Date;
  updatedAt: Date;
}

const EndpointSchema = new Schema<IEndpoint>(
  {
    projectId: { type: Schema.Types.ObjectId, ref: "Project", default: null, index: true },
    name: { type: String, required: true, trim: true },
    url: { type: String, required: true },
    method: {
      type: String,
      enum: ["GET", "POST", "PUT", "PATCH", "DELETE", "HEAD"],
      default: "GET",
    },
    expectedStatusCode: { type: Number, default: 200 },
    interval: { type: Number, default: 60, min: 60, max: 3600 },
    timeout: { type: Number, default: 10, min: 1, max: 60 },
    headers: { type: Map, of: String, default: {} },
    body: { type: String, default: "" },
    isActive: { type: Boolean, default: true },
    alertEnabled: { type: Boolean, default: false },
    alertThreshold: { type: Number, default: 3 },
    consecutiveFailures: { type: Number, default: 0 },
    isAlerting: { type: Boolean, default: false },
    lastAlertedAt: { type: Date, default: null },
    notifications: {
      email: {
        enabled: { type: Boolean, default: false },
        address: { type: String },
      },
      discord: {
        enabled: { type: Boolean, default: false },
        webhookUrl: { type: String },
      },
      webhook: {
        enabled: { type: Boolean, default: false },
        url: { type: String },
      },
    },
    lastCheckedAt: { type: Date, default: null },
    lastStatus: {
      type: String,
      enum: ["UP", "DOWN", "DEGRADED", null],
      default: null,
    },
    lastResponseTime: { type: Number, default: null },
    totalChecks: { type: Number, default: 0 },
    successfulChecks: { type: Number, default: 0 },
    uptimePercentage: { type: Number, default: 100 },
  },
  { timestamps: true }
);

EndpointSchema.index({ isActive: 1 });
EndpointSchema.index({ lastCheckedAt: 1 });

export default mongoose.models.Endpoint ||
  mongoose.model<IEndpoint>("Endpoint", EndpointSchema);

File: backend/models/endpoint.py
from datetime import datetime, timezone
from typing import Literal, Optional

from beanie import Document, PydanticObjectId
from pydantic import BaseModel, Field


class EmailNotification(BaseModel):
    enabled: bool = False
    address: Optional[str] = None


class DiscordNotification(BaseModel):
    enabled: bool = False
    webhook_url: Optional[str] = Field(default=None, alias="webhookUrl")

    model_config = {"populate_by_name": True}


class WebhookNotification(BaseModel):
    enabled: bool = False
    url: Optional[str] = None


class Notifications(BaseModel):
    email: EmailNotification = EmailNotification()
    discord: DiscordNotification = DiscordNotification()
    webhook: WebhookNotification = WebhookNotification()


class Endpoint(Document):
    project_id: Optional[PydanticObjectId] = Field(default=None, alias="projectId")
    name: str
    url: str
    method: Literal["GET", "POST", "PUT", "PATCH", "DELETE", "HEAD"] = "GET"
    expected_status_code: int = Field(default=200, alias="expectedStatusCode")
    interval: int = Field(default=60, ge=60, le=3600)
    timeout: int = Field(default=10, ge=1, le=60)
    headers: dict[str, str] = Field(default_factory=dict)
    body: str = ""
    is_active: bool = Field(default=True, alias="isActive")

    alert_enabled: bool = Field(default=False, alias="alertEnabled")
    alert_threshold: int = Field(default=3, alias="alertThreshold")
    consecutive_failures: int = Field(default=0, alias="consecutiveFailures")
    is_alerting: bool = Field(default=False, alias="isAlerting")
    last_alerted_at: Optional[datetime] = Field(default=None, alias="lastAlertedAt")

    notifications: Notifications = Notifications()

    last_checked_at: Optional[datetime] = Field(default=None, alias="lastCheckedAt")
    last_status: Optional[Literal["UP", "DOWN", "DEGRADED"]] = Field(default=None, alias="lastStatus")
    last_response_time: Optional[float] = Field(default=None, alias="lastResponseTime")
    total_checks: int = Field(default=0, alias="totalChecks")
    successful_checks: int = Field(default=0, alias="successfulChecks")
    uptime_percentage: float = Field(default=100.0, alias="uptimePercentage")

    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc), alias="createdAt")
    updated_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc), alias="updatedAt")

    class Settings:
        name = "endpoints"

    model_config = {"populate_by_name": True}
